src/Interpreters/Context.cpp

```diff
@@ -3447,4 +3448,14 @@ WriteSettings Context::getWriteSettings() const
     return res;
 }
 
+bool Context::isAccessAlreadyChecked() const
+{
+    return access_already_checked;
+}
+
+void Context::setAccessAlreadyChecked()
+{
+    access_already_checked = true;
```

Comment:
Currently there is not way to check access second time for ON CLUSTER query, because we do not pass any information about user. Checking access with global context when user is nullptr just does not make sense. We could write user name into DDL task in [Zoo]Keeper or something like that, but it will not work for inconsistent configurations (however, we do not provide guarantees that everything will work as expected for inconsistent configurations, so maybe it's ok)